### ISSUE TITLE
How to utilize the pre-calculated similarity matrix? 

### ISSUE THREAD

**Issue by U70-TK** [Mouradost/STGM#13]
Hi, I was trying to run the program, but it seems that if I run it without using pre-calculated similarity matrix, it simply doesn't work, 10 hours have passed and the first epoch still hasn't finished yet. 

So then I followed the instruction in README and calculated the similarity matrix using the provided dtw.py file with command `python ./src/helpers/dtw.py ~/Documents/Datasets/METRLA/metr-la.h5` And now I get the "dtw.npy" file, but when I re-run the whole program, the line "No pre-calculated contribution matrix was provided! ..." still pops out. 

I tried to set the matrix directly to self.sim in "datasets/base.py" on line 39 (which doesn't seem to be correct) using self.sim = np.load(). And the training loss and validation loss it gave me don't make any sense. 

Also just a little bit curious, I can see that in `__getItem(self, idx)__` function in "datasets/base.py", dtw() function is called each time the train_data is enumerated with different indexes, but by default the "dtw.npy" file only stores the similarity matrix `data[0:None, 0:None], data[0:None, 0:None]`. Can this "dtw.npy" file even be good to use if the index is different when enumerating? 

I'm getting stuck for a long time, any help is highly appreciated!!!

**Comment by Mouradost** [comment]
Hi, take a look at #10 maybe it's going to help you.

When running the dtw.py on the fly (without providing a pre-calculated sim matrix) it takes a very long-time, especially if the CPU is weak. Since the training loop has to calculate the sim matrix for each time window and for each node pairs $(n_i, n_j)$ at each step.

You do not need to modify the `base.py`, instead inherit from `base.py` and overwrite the `load_data` similar to the following.

```python
# ML
import pandas as pd
import numpy as np

# Own
from datasets.base import BaseDataset


class Dataset(BaseDataset):
    def load_data(self):
        self._name = "PEMSBAY"
        self.adj = pd.read_csv(
            self.data_folder_path / "PEMSBAY" / "W_pemsbay.csv"
        ).values
        data = pd.read_hdf(self.data_folder_path / "PEMSBAY" / "pems-bay.h5")
        self.timestamps = data.index
        self.timestamps.freq = self.timestamps.inferred_freq
        self.time_index = self.time_to_idx(self.timestamps, freq="5min")
        self.data = data.values
        self.data_mean, self.data_std = self.data.mean(), self.data.std()
        self.data_min, self.data_max = self.data.min(), self.data.max()
        # Do not forget to normalize the similarity matrix with the build-in function, this is important in order to have correct training.
        self.sim = self.normalize_sim(np.load('dtw.npy'))
```

For the current version of `dtw.py` it will calculate a global similarity matrix, I didn't have time to implement the par time window code which should produce a sim matrix with `(M, N, N)` where `N` is the number of nodes and `M` is the dataset size after windowing. I will release an update in the near future.

If this is too much work, you can first train **STGM** without the similarity estimator by running the following command.

```shell
./run.py model=STGM trainer=Default
``` 

If you do so set the `self.sim = torch.ones(N, N)` so that the dtw.py is not run during training since it's not needed.

The performance is close to STGM_FULL, and I will soon release the relevant code that allows training STGM and the Estimator separately, so you can use your pre-trained STGM.

**Comment by U70-TK** [comment]
Thank you so much for your quick response! Thats so helpful!

Just a quick question tho, could you please take a quick look at the W_metrla.csv file I used: [https://github.com/deepkashiwa20/DL-Traff-Graph/blob/main/METRLA/W_metrla.csv](https://github.com/deepkashiwa20/DL-Traff-Graph/blob/main/METRLA/W_metrla.csv) is this good to use or do I need to download the adj-METR-LA.pkl file you posted earlier and transfer it using the code below: 
```
import pickle
import pandas as pd 

file_path = "./raw_data/METR-LA/adj_METR-LA.pkl"
out_path = "./datacsv/W_metrla.csv"
with open(file_path, 'rb') as file:
    data = pickle.load(file,encoding='latin1')

data = data[2]
df = pd.DataFrame(data)
df.to_csv(out_path,index=False)
``` 
The values in the csv file generated by the above code looks pretty small, is it already normalized? Does it mean I need to get rid of normalize_adj() call in base.py? 

And Yeah, I will indeed try to follow your instruction in issues #10 and train them separately, but anyways I'm still looking forward to see your versions soon lol.

**Comment by Mouradost** [comment]
I took a look at the file provided in the link and it seems correct to me and no the file you provided is not normalized you still need to normalize it.

The `base.py` is not meant to be modified you can overwrite what ever you need just by inheriting from `BaseDataset`.

**Comment by U70-TK** [comment]
OK sounds good, thanks for your precise and quick response. Good luck with your research and have a wonderful day!

**Comment by Mouradost** [comment]
You're welcome. Thanks wish you the same
